Libraries/Opc.Ua.Security.Certificates/Extensions/X509SubjectAltNameExtension.cs

```diff
@@ -0,0 +1,446 @@
+                buffer.Append(m_ipAddresses[ii]);
+            }
+
+            return buffer.ToString();
+        }
+
+        /// <summary>
+        /// Initializes the extension from ASN.1 encoded data.
+        /// </summary>
+        public override void CopyFrom(AsnEncodedData asnEncodedData)
+        {
+            if (asnEncodedData == null) throw new ArgumentNullException(nameof(asnEncodedData));
+            Oid = asnEncodedData.Oid;
+            RawData = asnEncodedData.RawData;
+            m_decoded = false;
+        }
+        #endregion
+
+        #region Public Properties
+        /// <summary>
+        /// The OID for a Subject Alternate Name extension.
+        /// </summary>
+        public const string SubjectAltNameOid = "2.5.29.7";
+
+        /// <summary>
+        /// The OID for a Subject Alternate Name 2 extension.
+        /// </summary>
+        public static string SubjectAltName2Oid = "2.5.29.17";
+
+        /// <summary>
+        /// Gets the uris.
+        /// </summary>
+        /// <value>The uris.</value>
+        public IReadOnlyList<string> Uris
+        {
+            get
+            {
+                EnsureDecoded();
+                return m_uris.AsReadOnly();
+            }
+        }
+
+        /// <summary>
+        /// Gets the domain names.
+        /// </summary>
+        /// <value>The domain names.</value>
+        public IReadOnlyList<string> DomainNames
+        {
+            get
+            {
+                EnsureDecoded();
+                return m_domainNames.AsReadOnly();
+            }
+        }
+
+        /// <summary>
+        /// Gets the IP addresses.
+        /// </summary>
+        /// <value>The IP addresses.</value>
+        public IReadOnlyList<string> IPAddresses
+        {
+            get
+            {
+                EnsureDecoded();
+                return m_ipAddresses.AsReadOnly();
+            }
+        }
+        #endregion
+
+        #region Private Methods
+        /// <summary>
+        /// Create a normalized IPv4 or IPv6 address from a 4 byte or 16 byte array.
+        /// </summary>
+        private string IPAddressToString(byte[] encodedIPAddress)
+        {
+            try
+            {
+                IPAddress address = new IPAddress(encodedIPAddress);
+                return address.ToString();
+            }
+            catch
+            {
+                throw new CryptographicException("Certificate contains invalid IP address.");
+            }
+        }
+
+#if NETSTANDARD2_1
+        /// <summary>
+        /// Encode the Subject Alternative name extension.
+        /// </summary>
+        private byte[] Encode()
+        {
+            var sanBuilder = new SubjectAlternativeNameBuilder();
+            foreach (var uri in m_uris)
+            {
+                sanBuilder.AddUri(new Uri(uri));
+            }
+            EncodeGeneralNames(sanBuilder, m_domainNames);
+            EncodeGeneralNames(sanBuilder, m_ipAddresses);
```

Comment:
You're accepting a list that could be alternating between dnsName and ipAddress values, but you separate them and write them normalized as all dnsNames first.  I'm just pointing it out.

Reply:
ok, I will check it. For our application order is not important but I understand it were better to keep it.